lib/DDG/Rewrite.pm

```diff
@@ -135,53 +135,72 @@ sub _build_nginx_conf {
 	my $cfg = "location ^~ ".$self->path." {\n";
 	$cfg .= "\tproxy_set_header Accept '".$self->accept_header."';\n" if $self->accept_header;
 	
-        if($uses_echo_module) {
-            # we need to make sure we have plain text coming back until we have a way
-            # to unilaterally gunzip responses from the upstream since the echo module
-            # will intersperse plaintext with gzip which results in encoding errors.
-            # https://github.com/agentzh/echo-nginx-module/issues/30
-            $cfg .= "\tproxy_set_header Accept-Encoding '';\n";
-
-            # This is a workaround that deals with endpoints that don't support callback functions.
-            # So endpoints that don't support callback functions return a content-type of 'application/json'
-            # because what they're returning is not meant to be executed in the first place.
-            # Setting content-type to application/javascript for those endpoints solves blocking due to 
-            # mime type mismatches.
-            $cfg .= "\tmore_set_headers 'Content-Type: application/javascript; charset=utf-8';\n";
-        }
+	if($uses_echo_module) {
+		# we need to make sure we have plain text coming back until we have a way
+		# to unilaterally gunzip responses from the upstream since the echo module
+		# will intersperse plaintext with gzip which results in encoding errors.
+		# https://github.com/agentzh/echo-nginx-module/issues/30
+		$cfg .= "\tproxy_set_header Accept-Encoding '';\n";
+
+		# This is a workaround that deals with endpoints that don't support callback functions.
+		# So endpoints that don't support callback functions return a content-type of 'application/json'
+		# because what they're returning is not meant to be executed in the first place.
+		# Setting content-type to application/javascript for those endpoints solves blocking due to 
+		# mime type mismatches.
+		$cfg .= "\tmore_set_headers 'Content-Type: application/javascript; charset=utf-8';\n";
+	}
 
 	if($uses_echo_module || $self->accept_header || $is_duckduckgo) {
-	    $cfg .= "\tinclude /usr/local/nginx/conf/nginx_inc_proxy_headers.conf;\n";
+		$cfg .= "\tinclude /usr/local/nginx/conf/nginx_inc_proxy_headers.conf;\n";
 	}
 
 	$cfg .= "\techo_before_body '".$self->callback."(';\n" if $wrap_jsonp_callback;
 	$cfg .= "\techo_before_body '".$self->callback.qq|("';\n| if $wrap_string_callback;
+
+	my $spice_name;
+	if(($spice_name) = $self->path =~ m/js\/spice\/(.+)\/?/){
+		$spice_name =~ s/\/$//;
+		$spice_name =~ s|/|_|og;
+	}
+
+	my $upstream = '';
+	if (defined $spice_name) {
+		$upstream = '$'.$spice_name.'_upstream';
+		$cfg .= "\tset $upstream $scheme://$host:$port;\n";
+	}
+
 	$cfg .= "\trewrite ^".$self->path.($self->has_from ? $self->from : "(.*)")." ".$uri_path." break;\n";
-	$cfg .= "\tproxy_pass ".$scheme."://".$host.":".$port."/;\n";
+
+	if ($upstream) {
```

Comment:
I imagine goodies don't need this since they're not involved with api calls.  So I wonder why we even have this spice/not spice logic here.

With the `set` issue, can we move the entire `if/else` above the rewrite line?  Or do the `proxy_pass` directives need to come after it?  If they do, we should at least be able to collapse the two `if`s above and localize `$spice_name`.